### DISCUSSION TITLE
Teensy 4 and buttonmatrix

### DISCUSSION THREAD

**Discussion by antomexita** [tttapa/Control-Surface#760] [Q&A]
Hi to all,
i need to connect a matrix keyboard on a teensy 4.
Each button have a diode.
I make the connection and use this code
`#include <Control_Surface.h>
#include <AH/Hardware/MCP23017Encoders.hpp>
#include <Wire.h>
USBDebugMIDI_Interface midi;
const AddressMatrix<6, 9> addresses = {{
    {24, 30, 36, 42, 48, 54, 60, 66, 72},
    {25, 31, 37, 43, 49, 55, 61, 67},
    {26, 32, 38, 44, 50, 56, 62, 68},
    {27, 33, 39, 45, 51, 57, 63, 69},
    {28, 34, 40, 46, 52, 58, 64, 70},
    {29, 35, 41, 47, 53, 59, 65, 71},
  } 
};
NoteButtonMatrix<6, 9> buttonmatrix = {
  {5, 6, 7, 8, 9, 10},
  {0, 1, 2, 3, 4, 11, 12, 13, 14},
  addresses,    // address
  CHANNEL_1,    // channel and cable number
};
void setup() {
  Control_Surface.begin();
}
void loop() {
  Control_Surface.loop();
}`
When touching a random key, the response on the serial monitor is a fast and continuous Note On, Note off. Sometime some key work or stop the On Off after some seconds.  The hardware connection is ok, work fine with Arduino Mega. I check the state of the pins, the column pins is HI and the row pins is LOW. 
I resolve it with a 5kohm pullup resistor.  
The answer is, the internal pullup resistor is activate automatically? 
If the code and hardware work fine with arduino, is possible a voltage incompatibility on teensy when use a diode (3v vs 5v) 
that makes the teensy work on the indeterminate logic level zone ?

**Comment by tttapa** [comment]
For a pure software solution, you can try increasing the delay time here:

https://github.com/tttapa/Control-Surface/blob/4b6011518fe056b4bbc289628af9be47dd5cfe35/src/AH/Hardware/ButtonMatrix.ipp#L26

> the internal pullup resistor is activate automatically?

Yes: 

https://github.com/tttapa/Control-Surface/blob/4b6011518fe056b4bbc289628af9be47dd5cfe35/src/AH/Hardware/ButtonMatrix.ipp#L44-L47

> is possible a voltage incompatibility on teensy when use a diode (3v vs 5v)
that makes the teensy work on the indeterminate logic level zone ?

Unlikely, I expect that the issue is the capacitance of the wiring and buttons, combined with the high resistance of the built-in pull-ups of the Teensy, and its faster GPIO and CPU.

**Reply by antomexita** [reply]
Hi Pieter, resolve the problem removing the resistors and increase the delay at 10ms. 
Thanks for the support

**Comment by antomexita** [comment]
Hi to all,
I want to contribute to this post by pointing out that the same problem occurred with reading the values ​​of 74hct4067 under teensy 4. That is double and erroneous readings of the values. Open the AnalogMultiplex.hpp and changed the "delayMicroseconds" value from 1 to 10. Now it works perfectly.
¡thats all folks!
